### ISSUE TITLE
Sent commands get flash on screen but no effect.

### ISSUE THREAD

**Issue by corpr8** [alanreid/bravia#5]
I've got a Bravia UhD and this code pairs ok with it - but when it tries to send a command, I just get a flash on sony logo / on screen but no effect.

I also extended the makeCommandRequest with a response function but all I see coming back is a blank response.

Any suggestions?

**Comment by corpr8** [comment]
Looking at this a little more with Wireshark running and comparing against the Chrome advanced rest client (which works) - 

The working control packet has a section:

Authorisation: Basic OjA1Nzc=\r\n
    Credentials: 0527

Whilst the packet sent by this script has that part missing and gets a 403 Forbidden in response.

Any help greatly appreciated.

**Comment by alanreid** [comment]
Hi @corpr8,

That `OjA1Nzc=` decodes to `:0577` in base64, meaning no user and `0577` as password. Maybe it's sending other credentials than the ones you're inputing?

I wish I could be of more help, but unfortunately I moved and I don't have a Bravia TV anymore :( 

This code worked with the `KDL-50W805` model, but they might have changed the authentication protocol for the new UHD models. 

Let me know if you make it work!

**Comment by alanreid** [comment]
Maybe this is the same as this? https://github.com/alanreid/bravia/issues/4#issuecomment-167063289

**Comment by corpr8** [comment]
So I did have to set permissions on the cookiejar file - I can see that it is getting and saving a cookie - just doesn't seem to use them when it tries to send a command (Doesn't seem to use basic http authentication).

Will have a play and see where I get...

**Comment by corpr8** [comment]
Ok - so got it working!

In index.js - Bravia.prototype.request, I changed:
  options.jar = this.cookieJar;
to:
  options.jar = true;

also I added:

  options.auth = {
    'user': '',
    'pass': keyCode
  }

Where keyCode is a new global variable which stores the 4 digit code captured within the key request:

```
rl.question('Please enter the 4-digit code shown on your TV: ', function(challenge) {
  //only store the code if we don't already have one...
  if (keyCode.length <1){
  keyCode = challenge.toString();
  }
```

Hope this helps others!

**Comment by alanreid** [comment]
Awesome! Could you send me a PR or a patch of what you edited so I can include it?

**Comment by xuncia** [comment]
Hi @corpr8 can you please share your index.js file I'll be happy to see if it works with my bravia, right now I'm really running out of solutions.

**Comment by corpr8** [comment]
Hey so what I did in the end was to get the commands working directly in advanced soap client (chrome extension). From there, I wrote curl commands which generated the same traffic (as seen by wireshark) - and dropped them into batch files.

For instance, my power on script looks like this:

#!/bin/bash  
CURL='/usr/bin/curl'

$CURL -v -XPOST http://192.168.1.68/sony/IRCC -A "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/47.0.2526.106 Safari/537.36" --cookie "auth=324BC8B152B8507C09305CB669FB2980D09B84E4" --header 'SOAPACTION: "urn:schemas-sony-com:service:IRCC:1#X_SendIRCC"' --header 'Accept-Encoding: gzip, deflate' --header 'Accept-Language: en-GB,en-US;q=0.8,en;q=0.6' --header 'Connection: keep-alive' --header 'Content-Type:application/atom+xml;' --header 'Origin: chrome-extension://hgmloofddffdnphfgcellkdfbfbjeloo' -d '<?xml version="1.0" encoding="utf-8"?><s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/" s:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/"><s:Body><u:X_SendIRCC xmlns:u="urn:schemas-sony-com:service:IRCC:1"><IRCCCode>AAAAAQAAAAEAAAAVAw==</IRCCCode>/u:X_SendIRCC/s:Body/s:Envelope'

Where the auth is ripped from the chrome browser cookie and the origin is set to match the Advanced soap client. (You can get all off this from developer mode) From there, I integrated it with Amazon Echo and now have voice powered remote on / off for the TV :-)

What really helped me was checking the curl generated requests against the chrome plugin generated requests...

Hope this helps!

**Comment by xuncia** [comment]
thanks for sharing @corpr8 , I did the same. If anyone else has the same problem here there is a good guide how to do 

http://www.openremote.org/display/forums/Sony+TV+HTTP+control

if you need a wol command you can use also wolcmd.
You can also add some more commands for example "turn on channel 46" (my children favourite in Italy), calling a new accessory in the config.json channel 46 and use  a new bash that sending the correct post for that channels.
Hope that helps too.

**Comment by alanreid** [comment]
Hi there,

I got a Bravia TV for a week again and I was able to reproduce and fix the issue. 
The problem was with the tough-cookie package (`tough-cookie/lib/cookie.js:576`), which replaces the tailing slash only on lookup and not on save. This means it's not able to find a cookie that's saved as `/sony/`, because it will always look it up as `/sony`. 

So pull the latest changes and it should work again.

Hope it helps!
Alan
